packages/@lwc/engine-core/src/framework/base-bridge-element.ts

```diff
@@ -125,6 +132,43 @@ export function HTMLBridgeElementFactory(
         };
     }
     defineProperties(HTMLBridgeElement.prototype, descriptors);
+    // observable attributes
+    HTMLBridgeElement.prototype.attributeChangedCallback = function (
+        this: HTMLElement,
+        attrName: string,
+        oldValue: string,
+        newValue: string
+    ) {
+        if (oldValue === newValue) {
+            // Ignore same values.
+            return;
+        }
+        const propName = attributeToPropMap[attrName];
+        if (isUndefined(propName)) {
+            if (!isUndefined(superAttributeChangedCallback)) {
+                // delegate unknown attributes to the super.
+                superAttributeChangedCallback.apply(this, arguments);
+            }
+            return;
+        }
+        if (!isAttributeLocked(this, attrName)) {
+            // Ignore changes triggered by the engine itself during:
+            // * diffing when public props are attempting to reflect to the DOM
+            // * component via `this.setAttribute()`, should never update the prop
+            // Both cases, the setAttribute call is always wrapped by the unlocking of the
+            // attribute to be changed
+            return;
+        }
+        // Reflect attribute change to the corresponding property when changed from outside.
+        (this as any)[propName] = newValue;
+    };
+    // Specify attributes for which we want to reflect changes back to their corresponding
+    // properties via attributeChangedCallback.
+    defineProperty(HTMLBridgeElement, 'observedAttributes', {
+        get() {
```

Comment:
making this a read only getter helps to avoid the extra computation, it is only needed when attempting to register a class the extends this class :)